README.md
## Intro

`OSM Legal Speed` is a Java SDK which help determine the legal max speed for most of the countries of the world, with respect to regional legal restrictions, based on OSM tags.  

This project is highly inspired by [OSM Legal Default Speeds](https://github.com/westnordost/osm-legal-default-speeds). Under the hood, we are using the JSON file generated by its [parser](https://github.com/westnordost/osm-legal-default-speeds/tree/master/parser) (`legal_default_speeds.json`).

## Getting started

OSM Legal Speed is a complete package to determine the legal speed based on a set of GPS coordinates and a vehicle type. Also, its providing support for different situation, including time based speed restrictions.

```java
package com.chargetrip.example;

import com.chargetrip.osmLegalSpeed.types.VehicleType;
import com.chargetrip.osmLegalSpeed.types.Options;

import java.io.IOException;
import java.text.ParseException;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

public class Main {
  public static void main(String[] args) {
    Map<String, String> tags = new HashMap<>();
    tags.put("type", "route");
    tags.put("route", "road");
    tags.put("highway", "motorway");

    try {
      Options options = new Options();
      options.latitude = 52.3727598;
      options.longitude = 4.8936041;
      options.datetime = LocalDateTime.now().withHour(12).withMinute(0);

      LegalSpeed legalSpeed = new LegalSpeed();

      // How to get a country with region code
      String countryWithRegion = legalSpeed.getCountryWithRegion(options.latitude, options.longitude);
      System.out.println("Country with region: " + countryWithRegion);

      // How to search for max speed tags
      LegalSpeed.SearchResult searchResult = legalSpeed.searchSpeedLimits(tags, countryWithRegion, true);
      System.out.println("SearchResult: ");
      System.out.println("- certitude: " + searchResult.certitude);
      System.out.println("- tags: " + searchResult.speedType);

      // How to get the legal speed for all vehicles
      System.out.println("Legal speeds:");
      for (VehicleType vehicle : VehicleType.values()) {
        options.vehicle = vehicle;
        System.out.println("- " + vehicle + ": " + legalSpeed.getSpeedLimit(tags, options));
      }
    } catch (IOException | ParseException e) {
      e.printStackTrace();
    }
  }
}
```

This example will output:

```shell
Country with region: NL
SearchResult: 
- certitude: Exact
- tags: motorway: {maxspeed:motorhome:conditional=80 @ (maxweightrating>3.5), maxspeed:conditional=100 @ (06:00-19:00); 90 @ (trailer); 80 @ (maxweightrating>3.5 AND trailer), maxspeed:hgv=80, maxspeed=130, minspeed=60, maxspeed:bus=80, maxspeed:coach=100}
Legal speeds:
- Car: 100.0
- Bus: 80.0
- MiniBus: 100.0
- SchoolBus: 100.0
- TruckBus: 100.0
- Coach: 100.0
- Goods: 100.0
- Hazmat: 100.0
- Hgv: 80.0
- Motorcycle: 100.0
- Tricycle: 100.0
- MotorHome: 100.0
```

## Internals

The SDK is using `legal_default_speeds.json` rules to determine the legal max speed and `countries.json` to determine country and/or region (is necessary) based on GPS coordinates. The legal rules file was generated using this [parser](https://github.com/westnordost/osm-legal-default-speeds/tree/master/parser) based on this [wiki page](https://wiki.openstreetmap.org/wiki/Default_speed_limits). Once more rules will be added to this page or changed, we will parse it again and build a new set of rules.

The `countries.json` file contains definitions only for countries and regions which are present in `legal_default_speeds.json`. Every time we update `legal_default_speeds.json`, we also check to see if we need to adjust `countries.json`. 

## Alternative (which inspired us)

- [OSM Legal Default Speeds](https://github.com/westnordost/osm-legal-default-speeds) - Kotlin multiplatform library that provides information about legal default speed limits. Runs on JVM, native and JavaScript.

| SDK                      | OSM legal speed tags | Country with region | Speed based on GPS |
|--------------------------|----------------------|---------------------|--------------------|
| OSM Legal Speed          | ✓                    | ✓                   | ✓                  |
| OSM Legal Default Speeds | ✓                    | x                   | x                  |

## Sponsors

[![Chargetrip logo](https://chargetrip-files.s3.eu-central-1.amazonaws.com/logo-1.png)](https://www.chargetrip.com)


CHANGELOG.md
## [1.2.1-rc.1](https://github.com/chargetrip/osm-legal-speed/compare/1.2.0...1.2.1-rc.1) (2023-03-06)


### Bug Fixes

* Speed profile for GB ([94c5954](https://github.com/chargetrip/osm-legal-speed/commit/94c5954ea6b9845c6918f6759cf7247aca9ad9ef))

# [1.2.0](https://github.com/chargetrip/osm-legal-speed/compare/1.1.0...1.2.0) (2022-12-20)


### Features

* Use slf4j for logging instead of system ([48de022](https://github.com/chargetrip/osm-legal-speed/commit/48de0223ea6ffbafdc17479a83ceed8f19c1d414))

# [1.2.0-rc.1](https://github.com/chargetrip/osm-legal-speed/compare/1.1.0...1.2.0-rc.1) (2022-12-20)


### Features

* Use slf4j for logging instead of system ([48de022](https://github.com/chargetrip/osm-legal-speed/commit/48de0223ea6ffbafdc17479a83ceed8f19c1d414))

# [1.1.0](https://github.com/chargetrip/osm-legal-speed/compare/1.0.2...1.1.0) (2022-12-19)


### Bug Fixes

* BE borders and clean up OSM tags ([046970e](https://github.com/chargetrip/osm-legal-speed/commit/046970eb0ddaf0c52ebb70aa18af55dfdc34c44f))
* Broken parse expressions for date and time operation ([7ce1ca8](https://github.com/chargetrip/osm-legal-speed/commit/7ce1ca89a1d2e773af27784e92454d89c5d78181))
* Support for condition in maxspeed tag ([c8e2286](https://github.com/chargetrip/osm-legal-speed/commit/c8e228628763f51ab1e03af7cd1d31fb96cb041b))


### Features

* Germany no speed limit support and fix ArrayIndexOutOfBoundsException ([2c9d507](https://github.com/chargetrip/osm-legal-speed/commit/2c9d507bfc5bda781794b61a631fabe0284ad9a3))

# [1.1.0-rc.4](https://github.com/chargetrip/osm-legal-speed/compare/1.1.0-rc.3...1.1.0-rc.4) (2022-12-19)


### Bug Fixes

* BE borders and clean up OSM tags ([046970e](https://github.com/chargetrip/osm-legal-speed/commit/046970eb0ddaf0c52ebb70aa18af55dfdc34c44f))

# [1.1.0-rc.3](https://github.com/chargetrip/osm-legal-speed/compare/1.1.0-rc.2...1.1.0-rc.3) (2022-12-19)


### Bug Fixes

* Broken parse expressions for date and time operation ([7ce1ca8](https://github.com/chargetrip/osm-legal-speed/commit/7ce1ca89a1d2e773af27784e92454d89c5d78181))

# [1.1.0-rc.2](https://github.com/chargetrip/osm-legal-speed/compare/1.1.0-rc.1...1.1.0-rc.2) (2022-12-16)


### Bug Fixes

* Support for condition in maxspeed tag ([c8e2286](https://github.com/chargetrip/osm-legal-speed/commit/c8e228628763f51ab1e03af7cd1d31fb96cb041b))

# [1.1.0-rc.1](https://github.com/chargetrip/osm-legal-speed/compare/1.0.2...1.1.0-rc.1) (2022-12-16)


### Features

* Germany no speed limit support and fix ArrayIndexOutOfBoundsException ([2c9d507](https://github.com/chargetrip/osm-legal-speed/commit/2c9d507bfc5bda781794b61a631fabe0284ad9a3))

## [1.0.2](https://github.com/chargetrip/osm-legal-speed/compare/1.0.1...1.0.2) (2022-11-25)


### Bug Fixes

* Added missing tests ([8eb9543](https://github.com/chargetrip/osm-legal-speed/commit/8eb95432e8c2fa6e37fd3fbfa949334c6e2ca2f3))
* Ignore OSM maxspeed if cannot get value from it ([87c06e8](https://github.com/chargetrip/osm-legal-speed/commit/87c06e8ea6517773a393d4c59c3f66ad6c3cb911))
* Walking speed and maxspeed with country rules ([5416dcf](https://github.com/chargetrip/osm-legal-speed/commit/5416dcf8e05ada2bda03392e47c83ebac6e99fae))

## [1.0.2-rc.2](https://github.com/chargetrip/osm-legal-speed/compare/1.0.2-rc.1...1.0.2-rc.2) (2022-11-25)


### Bug Fixes

* Ignore OSM maxspeed if cannot get value from it ([87c06e8](https://github.com/chargetrip/osm-legal-speed/commit/87c06e8ea6517773a393d4c59c3f66ad6c3cb911))

## [1.0.2-rc.1](https://github.com/chargetrip/osm-legal-speed/compare/1.0.1...1.0.2-rc.1) (2022-11-24)


### Bug Fixes

* Walking speed and maxspeed with country rules ([5416dcf](https://github.com/chargetrip/osm-legal-speed/commit/5416dcf8e05ada2bda03392e47c83ebac6e99fae))

## [1.0.1](https://github.com/chargetrip/osm-legal-speed/compare/1.0.0...1.0.1) (2022-11-22)


### Bug Fixes

* Added missing doc blocks ([7dcbfad](https://github.com/chargetrip/osm-legal-speed/commit/7dcbfadc2b8cc5ad00167c82903da8a82bf47e38))
* **ci:** Put back maven GHA ([b6cbce2](https://github.com/chargetrip/osm-legal-speed/commit/b6cbce26b7127642e6cbfb9f414d777103b04041))
* **ci:** Update GHA publish action ([3bac9bd](https://github.com/chargetrip/osm-legal-speed/commit/3bac9bdd64077bbf76cf1a3ca26df704ef38ec5e))
* **ci:** Update GHA workflow ([2a3bdf8](https://github.com/chargetrip/osm-legal-speed/commit/2a3bdf83a7f8cbcdc445be139e28e08c25fc20c6))
* **ci:** Use  for packaging ([9c2cca6](https://github.com/chargetrip/osm-legal-speed/commit/9c2cca6f8fde0a7290dd623eb176f69b9ff0db82))
* Moved resources and fix number parse ; ([824a667](https://github.com/chargetrip/osm-legal-speed/commit/824a667fc33c17581b0693845139b68c68004048))
* **test:** Make sure tests run on UTC timezone ([a3128ed](https://github.com/chargetrip/osm-legal-speed/commit/a3128ed2b18ec50620e6e8e16685dd02939bbd72))

## [1.0.1-rc.6](https://github.com/chargetrip/osm-legal-speed/compare/1.0.1-rc.5...1.0.1-rc.6) (2022-11-22)


### Bug Fixes

* Moved resources and fix number parse ; ([824a667](https://github.com/chargetrip/osm-legal-speed/commit/824a667fc33c17581b0693845139b68c68004048))

## [1.0.1-rc.5](https://github.com/chargetrip/osm-legal-speed/compare/1.0.1-rc.4...1.0.1-rc.5) (2022-11-21)


### Bug Fixes

* Added missing doc blocks ([7dcbfad](https://github.com/chargetrip/osm-legal-speed/commit/7dcbfadc2b8cc5ad00167c82903da8a82bf47e38))

## [1.0.1-rc.4](https://github.com/chargetrip/osm-legal-speed/compare/1.0.1-rc.3...1.0.1-rc.4) (2022-11-21)


### Bug Fixes

* **ci:** Put back maven GHA ([b6cbce2](https://github.com/chargetrip/osm-legal-speed/commit/b6cbce26b7127642e6cbfb9f414d777103b04041))
* **test:** Make sure tests run on UTC timezone ([a3128ed](https://github.com/chargetrip/osm-legal-speed/commit/a3128ed2b18ec50620e6e8e16685dd02939bbd72))

## [1.0.1-rc.3](https://github.com/chargetrip/osm-legal-speed/compare/1.0.1-rc.2...1.0.1-rc.3) (2022-11-21)


### Bug Fixes


src/main/java/com/chargetrip/osmLegalSpeed/Main.java
package com.chargetrip.osmLegalSpeed;

import com.chargetrip.osmLegalSpeed.types.Options;
import com.chargetrip.osmLegalSpeed.types.VehicleType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.simple.SimpleLogger;

import java.io.IOException;
import java.text.ParseException;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

/**
 * Example class on how the SDK works
 */
public class Main {
    /**
     * Main method as an example
     *
     * @param args Input arguments
     */
    public static void main(String[] args) {
        System.setProperty(SimpleLogger.DEFAULT_LOG_LEVEL_KEY, "TRACE");

        Logger logger = LoggerFactory.getLogger(Main.class);

        Map<String, String> tags = new HashMap<>();
        tags.put("type", "route");
        tags.put("route", "road");
        tags.put("highway", "motorway");

        try {
            Options options = new Options();
            options.latitude = 52.3727598;
            options.longitude = 4.8936041;
            options.datetime = LocalDateTime.now().withHour(12).withMinute(0);

            LegalSpeed legalSpeed = new LegalSpeed();

            // How to get a country with region code
            String countryWithRegion = legalSpeed.getCountryWithRegion(options.latitude, options.longitude);
            logger.debug("Country with region: " + countryWithRegion);

            // How to search for max speed tags
            LegalSpeed.SearchResult searchResult = legalSpeed.searchSpeedLimits(tags, countryWithRegion, true);
            logger.debug("SearchResult: ");
            logger.debug("- certitude: " + searchResult.certitude);
            logger.debug("- tags: " + searchResult.speedType);

            // How to get the legal speed for all vehicles
            logger.debug("Legal speeds:");
            for (VehicleType vehicle : VehicleType.values()) {
                options.vehicle = vehicle;
                logger.debug("- " + vehicle + ": " + legalSpeed.getSpeedLimit(tags, options));
            }

            // How to get speed from maxspeed tag
            options.vehicle = VehicleType.Car;
            tags.put("maxspeed", "50;30");
            logger.debug("From maxspeed with multiple values: " + legalSpeed.getSpeedLimit(tags, options));

            tags.put("maxspeed", "NL:urban");
            logger.debug("From maxspeed with parent rule: " + legalSpeed.getSpeedLimit(tags, options));

            tags.put("maxspeed", "null");
            logger.debug("From null maxspeed: " + legalSpeed.getSpeedLimit(tags, options));

            tags.put("maxspeed", "NL:trunk");
            logger.debug("From maxspeed with parent rule: " + legalSpeed.getSpeedLimit(tags, options));

            tags.put("maxspeed", "none");
            options.latitude = 52.5170365;
            options.longitude = 13.3888599;
            logger.debug("From none maxspeed in DE: " + legalSpeed.getSpeedLimit(tags, options));

            Map<String, String> germanyTags = new HashMap<>();
            germanyTags.put("destination:ref", "B 173");
            germanyTags.put("destination:lanes", "Dresden|Freiberg;Freital;Kesselsdorf");
            germanyTags.put("bdouble", "yes");
            germanyTags.put("oneway", "yes");
            germanyTags.put("turn:lanes", "left|right");
            germanyTags.put("lit", "no");
            germanyTags.put("hazmat", "designated");
            germanyTags.put("lanes", "2");
            germanyTags.put("highway", "motorway_link");
            logger.debug("From motorway_link in DE: " + legalSpeed.getSpeedLimit(germanyTags, options));

            Map<String, String> gbTags = new HashMap<>();
            gbTags.put("source:name", "OS_OpenData_Locator");
            gbTags.put("name", "Springfield Avenue");
            gbTags.put("highway", "residential");
            gbTags.put("oneway", "yes");
            options.latitude = 52.0486164;
            options.longitude = -2.7239301;

            logger.debug("From residential in GB: " + legalSpeed.getSpeedLimit(gbTags, options));
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
    }
}
